pkg/discovery/dtofactory/general_builder.go

```diff
@@ -381,16 +390,64 @@ func aggregateContainerThrottlingSamples(entityID string, samples []metrics.Thro
 	sort.SliceStable(samples, func(i, j int) bool {
 		return samples[i].Timestamp < samples[j].Timestamp
 	})
-	throttled = (samples[numberOfSamples-1].Throttled - samples[0].Throttled)
-	total = samples[numberOfSamples-1].Total - samples[0].Total
 
+	resetIndexes := []int{}
+	// calculate peak
 	for i := 0; i < numberOfSamples-1; i++ {
-		total := samples[i+1].Total - samples[i].Total
+		if samples[i+1].Total < samples[i].Total || samples[i+1].Throttled < samples[i].Throttled {
+			// This probably means the counter was reset for some reason
+			// we ignore this sample diff for our calculations
```

Comment:
@irfanurrehman Another really creative idea that takes advantage of the higher sampling rate to handle counter resets without needing to know the counter max!

In terms of calculating the average, the current optimized implementation consists of two loops and maintains the resetIndexes slice, kind of a little complex.  I wonder if it'd make sense to trade that off with a simplified implementation to use the first loop and accumulate the average within that as well, eliminating the second loop as well as the resetIndexes.  I understand that in most cases the 2nd loop will actually not be executed, but the introduced code complexity seems to me overweight the performance benefit - IMHO one extra add operation in each iteration is very minimum performance-wise so I'd vote for the simplicity.

Btw, about the frequency of counter resets, if I calculate it correctly, one will only happen when a container is up for 13 years - I wonder if one would ever live that long!

Reply:
@chlam4 you are awesome.. :-) Updated!

Reply:
Wow, very clever implementation even in terms of calculating the average.  Kudos!